Issue/document podman deployment (#7025)
inmanta/inmanta-core: docs/install/2-install-server-with-podman-and-systemd.rst

    please :ref:`jump to the end of this document<install-postgresql-with-podman>`, where we explain to you how to easily deploy a database using Postman and Systemd.

Pull the image
##############

.. only:: oss

    Use ``podman pull`` to get the desired image:

    .. code-block:: sh

        podman pull ghcr.io/inmanta/orchestrator:latest

    This command will pull the latest version of the Inmanta OSS Orchestrator image.

.. only:: iso

    Step 1: Log in to container registry
    -------------------------------------

    Connect to the container registry using your entitlement token.

    .. code-block:: console

        $ podman login containers.inmanta.com
        Username: containers
        Password: <your-entitlement-token>

        Login Succeeded
        $

    Replace ``<your-entitlement-token>`` with the entitlement token provided with your license.

    Step 2: Pull the image
    ----------------------

    Use ``podman pull`` to get the desired image:

    .. code-block:: sh
        :substitutions:

        podman pull containers.inmanta.com/containers/service-orchestrator:|version_major|

>     This command will pull the latest version of the Inmanta Service Orchestrator image.

Reviewer: For the ISO docs build, the tag is a major version number. So this statement is only true within the major version this documentation build is for.

Author: Is it better now?

    As of now, the container can not be configured with environment variables, we should use a configuration file, mounted inside of the container.
    To do this, you can get the current configuration file from the container, edit it, and mount it where it should be in the container.
    Let's create a file on the host at ``~/.config/inmanta/inmanta.cfg``. We can take as template the default file already packaged in our
    container image.

    .. only:: oss

        .. code-block:: sh

            mkdir -p ~/.config/inmanta
            podman run --rm -ti ghcr.io/inmanta/orchestrator:latest cat /etc/inmanta/inmanta.cfg > ~/.config/inmanta/inmanta.cfg

    .. only:: iso

        .. code-block:: sh
            :substitutions:

            mkdir -p ~/.config/inmanta
            podman run --rm -ti containers.inmanta.com/containers/service-orchestrator:|version_major| cat /etc/inmanta/inmanta.cfg > ~/.config/inmanta/inmanta.cfg

2.  Update database settings:
    It is very unlikely that your database setup will match the one described in the default config we just got.  Update the configuration in the ``[database]`` section
    to reflect the setup you have.

.. only:: iso

>     3.  Get the license files

Reviewer: ```suggestion
    3.  Get the license files:
```

>         Together with the access to the inmanta container repo, you should also have received a license and an entitlement files.  The orchestrator will need them

Reviewer: ```suggestion
        Together with the access to the inmanta container repo, you should also have received a license and an entitlement file. The orchestrator will need them
```

        .. code-block:: console

            $ tree .config/inmanta
            .config/inmanta
            ├── inmanta.cfg
            └── license
                ├── com.inmanta.jwe
                └── com.inmanta.license

            2 directories, 3 files

.. _setup-systemd-unit:

Start the server with systemd
#############################

Here is a systemd unit file that can be used to deploy the server on your machine.

.. only:: oss

    .. code-block:: 

        [Unit]
        Description=Podman 
        Documentation=https://docs.inmanta.com
        Wants=network-online.target
        After=network-online.target
        RequiresMountsFor=%t/containers

        [Service]
        Environment=PODMAN_SYSTEMD_UNIT=%n
        Restart=on-failure
        TimeoutStopSec=70
        ExecStart=/usr/bin/podman run \
                --cidfile=%t/%n.ctr-id \
                --cgroups=no-conmon \
                --sdnotify=conmon \
                -d \
                --replace \
                --publish=127.0.0.1:8888:8888 \
                --uidmap=993:@%U \
                --uidmap=0:1:993 \

>                 --gidmap=993:@%G \

Reviewer: The `@%U` and `@%G` syntax is not supported on the podman version available on Rocky Linux 8 and 9 as discussed.

Author: Fixed in https://github.com/inmanta/inmanta-core/pull/7025/commits/d4acb57193b6d0a51866e7ca6defe92276bfbaa1

        Documentation=https://docs.inmanta.com
        Wants=network-online.target
        After=network-online.target
        RequiresMountsFor=%t/containers

        [Service]
        Environment=PODMAN_SYSTEMD_UNIT=%n
        Restart=on-failure
        TimeoutStopSec=70
        ExecStart=/usr/bin/podman run \
                --cidfile=%t/%n.ctr-id \
                --cgroups=no-conmon \
                --sdnotify=conmon \
                -d \
                --replace \
                --publish=127.0.0.1:8888:8888 \
                --uidmap=993:0:1 \
                --uidmap=0:1:993 \
                --gidmap=993:0:1 \
                --gidmap=0:1:993 \
                --name=inmanta-orchestrator-server \
                --volume=%E/inmanta/inmanta.cfg:/etc/inmanta/inmanta.cfg:z \
                --volume=%E/inmanta/license/com.inmanta.license:/etc/inmanta/license/com.inmanta.license:z \
                --volume=%E/inmanta/license/com.inmanta.jwe:/etc/inmanta/license/com.inmanta.jwe:z \
                --entrypoint=/usr/bin/inmanta \
                --user=993:993 \
                containers.inmanta.com/containers/service-orchestrator:|version_major| \

>                 -vvv --timed-logs server

Reviewer: Like this the server logs will be sent to stdout and not to a file. Is this something we want? The RPM executes the following command to send the logs file: https://github.com/inmanta/inmanta-core/blob/ae61aa04b4b53447585a8ef055f5bb874dc6436e/misc/inmanta-server.service#L10

The same is true for the OSS service file of course.

Author: Yes, but that only works if the user didn't change the log folder configuration.  Or rather, it would always work, but be confusing if this log folder is not meant to be used.
I am not sure what is the best, I am open to both.

Reviewer: As discussed we will go for the approach were:

* We set ` --log-file /var/log/inmanta/server.log`. That way the support archive also captures the server logs.
* And, we map the `/var/log/inmanta` directory by default to the hosts to make the logs accessible for the server in case the server fails to start.

                --env=POSTGRES_PASSWORD=inmanta \
                docker.io/library/postgres:13 
        ExecStop=/usr/bin/podman stop \
                --ignore -t 10 \
                --cidfile=%t/%n.ctr-id
        ExecStopPost=/usr/bin/podman rm \
                -f \
                --ignore -t 10 \
                --cidfile=%t/%n.ctr-id
        Type=notify
        NotifyAccess=all

        [Install]
        WantedBy=default.target

4.  Create the folder that will container the persistent storage for the database: ``~/.local/shared/inmanta-orchestrator-db/data``.

    .. code-block:: sh

        mkdir -p ~/.local/share/inmanta-orchestrator-db/data

5.  Reload the systemd daemon, enable the service, and start it.

    .. code-block:: sh

        systemctl --user daemon-reload
        systemctl --user enable inmanta-orchestrator-db.service
        systemctl --user start inmanta-orchestrator-db.service

6.  In the unit file of the orchestrator (as described :ref:`here<setup-systemd-unit>`), make sure to attach the orchestrator
    container to the network the database is a part of, using the ``--network`` option of the ``podman run`` command.

> 7.  Don't forget to update the ip address of the database in the inmanta server configuration file (``~/.config/inmanta/inmanta.cfg``)!

Author: Well, it will not be running yet, as it didn't have a database (I assumed that this setup would be done before the orchestrator setup)

.. _install-server-with-podman:

Install Inmanta with Podman and Systemd
***************************************

This page explains how to setup an Inmanta orchestration server using Podman and Systemd.

> This guide assumes you already have `Podman <http://podman.io/>`_ installed on your machine and that you are running a Linux distribution with Systemd.

Reviewer 2: ```suggestion
This guide assumes you already have `Podman <http://podman.io/>`_ installed on your machine and that you are running a Linux distribution using Systemd.
```

.. note::

>     The full setup should be doable without any root privilege (rootless) on the host, running the orchestrator with an unprivileged user.  

Reviewer 2: ```suggestion
    The full setup can be done root privileges (rootless) on the host and running the orchestrator with an unprivileged user.  
```

Author: I think a word is missing there, I am not sure of the point you are making

Reviewer 2: Like this. What I changed was the your version is very conditional: "should be doable" sounds like: we have not tried it but we so no reason why it would not work

Author: I formulated it completely differently, can you check again?

Pull the image
##############

.. only:: oss

    Use ``podman pull`` to get the desired image:

    .. code-block:: sh

        podman pull ghcr.io/inmanta/orchestrator:latest

    This command will pull the latest version of the Inmanta OSS Orchestrator image.

.. only:: iso

    Step 1: Log in to container registry
    -------------------------------------

    Connect to the container registry using your entitlement token.

    .. code-block:: console

        $ podman login containers.inmanta.com
        Username: containers
        Password: <your-entitlement-token>

        Login Succeeded
        $

    Replace ``<your-entitlement-token>`` with the entitlement token provided with your license.

    Step 2: Pull the image
    ----------------------

    Use ``podman pull`` to get the desired image:

    .. code-block:: sh
        :substitutions:

        podman pull containers.inmanta.com/containers/service-orchestrator:|version_major|

    This command will pull the latest release of the Inmanta Service Orchestrator image within this major version.

Prepare the orchestrator configuration
######################################

1.  Get the default configuration file:

>     As of now, the container can not be configured with environment variables, we should use a configuration file, mounted inside of the container.

Reviewer 2: ```suggestion
    As of now, the container cannot be configured with environment variables, we should use a configuration file, mounted inside the container.
```

            podman run --rm -ti ghcr.io/inmanta/orchestrator:latest cat /etc/inmanta/inmanta.cfg > ~/.config/inmanta/inmanta.cfg

    .. only:: iso

        .. code-block:: sh
            :substitutions:

            mkdir -p ~/.config/inmanta
            podman run --rm -ti containers.inmanta.com/containers/service-orchestrator:|version_major| cat /etc/inmanta/inmanta.cfg > ~/.config/inmanta/inmanta.cfg

2.  Update database settings:
    It is very unlikely that your database setup will match the one described in the default config we just got.  Update the configuration in the ``[database]`` section
    to reflect the setup you have.

3.  Make sure that there is a folder on your host that can persist all the logs of the server and that it is owned by the user running the orchestrator service.  
    In this setup, the folder on the host will be ``~/.local/share/inmanta-orchestrator-server/logs``.

    .. code-block:: sh

        mkdir -p ~/.local/share/inmanta-orchestrator-server/logs

    .. warning:: 
        Inside of the container, this folder will be mounted at ``/var/log/inmanta`` as it is the default location where the orchestrator saves its logs.  This

>         location is configurable in the orchestrator configuration file.  If you were, for any reason, to change this location, make sure to update any usage

Reviewer 2: ```suggestion
        location is configurable in the orchestrator configuration file.  If you for any reason would change this location in the mount, make sure to update any usage
```

.. only:: iso

    4.  Get the license files:
        Together with the access to the inmanta container repo, you should also have received a license and an entitlement file. The orchestrator will need them
        in order to run properly.  You can also place them in a config directory on your host.  After this step, we assume that this folder is
        ``~/.config/inmanta/license/`` and that both files are named ``com.inmanta.license`` and ``com.inmanta.jwe`` respectively.

        .. code-block:: console

            $ tree .config/inmanta
            .config/inmanta
            ├── inmanta.cfg
            └── license
                ├── com.inmanta.jwe
                └── com.inmanta.license

            2 directories, 3 files

.. _setup-systemd-unit:

Start the server with systemd
#############################

Here is a systemd unit file that can be used to deploy the server on your machine.

.. only:: oss

    .. code-block:: 

        [Unit]
        Description=Podman 
        Documentation=https://docs.inmanta.com
        Wants=network-online.target
        After=network-online.target
        RequiresMountsFor=%t/containers

>         [Service]

Reviewer 2: I would also like to have the option where the user that is setting this up has root permissions (either working as root or working as the system user using sudo). This would require a user and group setting here as well.

Author: I added a tab-set with rootful and rootless versions, I think it is fairly complete

        Environment=PODMAN_SYSTEMD_UNIT=%n
        Restart=on-failure
        TimeoutStopSec=70
        ExecStart=/usr/bin/podman run \
                --cidfile=%t/%n.ctr-id \
                --cgroups=no-conmon \
                --sdnotify=conmon \
                -d \
                --replace \
                --publish=127.0.0.1:8888:8888 \

>                 --uidmap=993:0:1 \

Reviewer 2: I would like a note that explains what these numbers are. Something like:

993 is the uid and gid of the inmanta user and group inside the container. This option instructs podman to map this user to our local non-root user.

Author: I added an explanation, is it clear enough or maybe too detailed?

                        --user=993:993 \
                        containers.inmanta.com/containers/service-orchestrator:|version_major| \
                        --log-file /var/log/inmanta/server.log --log-file-level 2 --timed-logs server
                ExecStop=/usr/bin/podman stop \
                        --ignore -t 10 \
                        --cidfile=%t/%n.ctr-id
                ExecStopPost=/usr/bin/podman rm \
                        -f \
                        --ignore -t 10 \
                        --cidfile=%t/%n.ctr-id
                Type=notify
                NotifyAccess=all

                [Install]
                WantedBy=default.target

        You can paste this configuration in a file named ``inmanta-orchestrator-server.service`` in the systemd folder for your user.
        This folder is typically ``~/.config/systemd/user/``.

    .. tab-item:: Rootful setup
        :sync: rootful-setup

        .. only:: oss

            .. code-block:: 

                [Unit]
                Description=Podman 
                Documentation=https://docs.inmanta.com
                Wants=network-online.target
                After=network-online.target

>                 RequiresMountsFor=%t/containers

Author: @bartv this will not work as the user of the unit is root and not inmanta.  I am not sure how to work around this, should we have two unit files? The root one starting the inmanta one? And the inmanta one starting the container?

Author: Same issue for `--cidfile=%t/%n.ctr-id`

Author: My question is rather: what is the sys-admin way of doing this?
